blocks/bricks/sequence_generators.py

```diff
@@ -20,49 +49,72 @@ class BaseSequenceGenerator(Initializable):
 
     This class combines two components, a readout network and an
     attention-equipped recurrent transition, into a context-dependent
-    sequence generator. Optionally a third component can be used which
+    sequence generator. Third component must be also given which
     forks feedback from the readout network to obtain inputs for the
     transition.
 
-    **Definitions:**
+    The class provides two methods: :meth:`generate` and :meth:`cost`. The
+    former is to actually generate sequences and the latter is to compute
+    the cost of generating given sequences.
 
-    * *States* of the generator are the states of the transition as
-      specified in `transition.state_names`.
+    The generation algorithm description follows.
 
-    * *Contexts* of the generator are the contexts of the transition as
-      specified in `transition.context_names`.
+    **Definitions and notation:**
 
-    * *Glimpses* are intermediate entities computed at every generation
-      step from states, contexts and the previous step glimpses. They are
-      computed in the transition's `apply` method when not given or by
-      explicitly calling the transition's `take_glimpses` method. The set
-      of glimpses considered is specified in `transition.glimpse_names`.
+    * States :math:`s_i` of the generator are the states of the transition
+      as specified in `transition.state_names`.
 
-    The generation algorithm description follows.
+    * Contexts of the generator are the contexts of the
+      transition as specified in `transition.context_names`.
+
+    * Glimpses :math:`g_i` are intermediate entities computed at every
+      generation step from states, contexts and the previous step glimpses.
+      They are computed in the transition's `apply` method when not given
+      or by explicitly calling the transition's `take_glimpses` method. The
+      set of glimpses considered is specified in
+      `transition.glimpse_names`.
+
+    * Outputs :math:`y_i` are produced at every step and form the output
+      sequence. A generation cost :math:`c_i` is assigned to each output.
 
     **Algorithm:**
 
-    1. The initial states are computed from the contexts. This includes
-       fake initial outputs given by the `initial_outputs` method of the
-       readout, initial states and glimpses given by the `initial_state`
-       method of the transition.
+    1. Initialization.
+
+       :math:`y_0 = readout.initial\_outputs(contexts)`
+
+       :math:`s_0, g_0 = transition.initial\_states(contexts)`
```

Comment:
Question about this:

It seems that s_0 is special-cased in that all the other s_i for i>0 depend on the (forked) feedback f_i, but s_0 doesn't depend on the f_0 which is computed from y_0. Is that the intention?

I was also wondering why in fact why we need a fixed "initial output" y_0 at all -- is the idea that this is used for a mandatory start-of-sequence marker for example? It seems like the only thing it influences is the first readout r_1. I'm guessing some of the more advanced models need it, but in a lot of cases it may be sufficient to specify the initial state s_0 to achieve desired control over the initial readout. And the fact that the y_0 is special-cased in that it doesn't influence s_0 makes it feel a bit awkward at first glance.